InterruptHandling.md

```diff
@@ -1,146 +1,400 @@
 ```
 
-The order of the registers is not really important when pushing them, but what is important is that we pop them we do in reverse order.
+In this example we stored `idtr` on the stack, which gets cleaned up when the function returns. This is okay because the IDTR register is like a segment register in that it caches whatever value was loaded into it. So it's okay that our idtr structure is no present, as the register will still have it. The actual IDT can't be on the stack, as the cpu does not cache that.
+
+At this point you should be able to install an interrupt handler into your IDT, load the IDT and set the interrupts flag. Your kernel will likely crash as soon as an interrupt is triggered though, as there are some special things we need to perform inside of an interrupt handler.
+
+## Interrupt Handler Stub
+
+Since an interrupt handler uses the same general purpose registers as the code that was interupted, we'll need to save and then restore the values of those registers, otherwise we may crash the interrupted program.
+There are a number of places you could store the state of these registers, we're going to use the stack as it's extremely simple to implement. In protected mode we have the `pusha`/`popa` instructions for this, but they're not present in long mode so we have to do this ourselves.
+
+There is also one other thing: when an interrupt is served the cpu will store some things on the stack, so that when the handler is done we can return to the previous code. The cpu pushes the following on to the stack (in this order):
+
+- `%ss`: The previous stack selector.
+- `%rsp`: The previous stack-top.
+- `%rflags`: The previous value of the flags register, before the cpu modified any flags for serving the interrupt.
+- `%cs`: The previous code selector.
+- `%rip`: The previous instruction pointer.
+
+Optionally, for some vectors the cpu will push a 64-bit error code (see the table below for specifics).
+This structure is known as an *iret frame*, because to return from an interrupt we use the `iret` instruction, which pops those five values from the stack. Hopefully the flow of things is clear at this point: the cpu serves the interrupt, pushes those five values onto the stack. Our handler function runs, and then executes the `iret` instruction to pop the previously pushed values off the stack, and return to the interrupted code.
+
+### An Example Stub
+
+Armed with the above infomation, you should be able to implement your own handler stubs. One common way to do this is using an assembler macro. Here you would create one macro that pushes all registers, calls a C function and then pops all registers before executing `iret`. What about the optional error code? Well, the easiest solution is to define *two* macros, one like the previous one, and another that pushes a pretend error code of 0, before pushing all the general registers. Because we know which vectors push an error code and which don't, we can change which macro we use. 
```

Comment:
I'm not sure what there would be to explain? You simply can't write that kind of code in C. Inline assembly does not guarantee the compiler won't interfere with the instructions (it may reorder them, or add extras). The only way to write an interrupt handler in C is the use the interrupt attribute, but then you can't access the saved registers, and I dont really want to suggest that.